README.md
# RSM

Visual stimulation software written by Bill Vinje 2012-2015
Uses MGL from Justin Gardner
Built to emulate lisp code visual stimulation from Chichilnisky lab used 1998-2015


Documentation/Changelog.txt

2014-09-25: BV just fixed the line colors in the part of the setup guide that talks about connecting Thuy's cable to the input line of the 6501. I believe that the reason for the original glitch is that Thuy and I apparently use a different polarity for the 6501 input line. That means that the current documentation applies to the cables made by Thuy (which should be all of them except for my initial testing cable). 

2014-09-25: BV Note that the call to "version -java" setup guide has a whitespace (should exaggerate in doc)
should return: "Java 1.7.0_11-b21 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode".

2014-09-25: BV The reason we are using java is because it lies underneath matlab. It forms part of matlab's infrastructure. 

2014-09-25: BV The video card referred to in the documentation is the ATI RADEON 5770. (I should make that explicit). This is probably the same as in the other desktop mac's that we have tested.  RSM also works on my laptop which has video card NVIDIA GeForce GT 750M.

2014-09-25.  EJC testing gets RSM working on a Mavericks machine, consistent with Bill’s experience that it works on laptop with Mavericks.  Video card not noted yet but it has one DVI and two thunderbolt connectors on it, works to first order at least.  Note that RSM forces 120Hz that monitor can handle, even though Displays preferences fails to do so.  This is good.  Plan is to set up a shared current version on DropBox soon for co-development with Greg.
 
2014-10-11.  BV (via EJC).  The normalization in line 19 of the gun values needs to be uncommented for the interpolated gamma tables to work properly.  It is strange that the line should have been commented out. An error of omission would be one thing, but it turns out this was commented out for a reason.   For right now: uncomment line 19 in Interpolate_Radpow_Data.m and use the interpolated table option (which is the default option). In my experience these interpolated tables have typically proven slightly better than the best-fit tables at giving good linearity at the high luminance end of the gray-scale linearity testing.


2014-10-17.  XY

•	Moving_Bar.m
o	Enable the bar to move in arbitrary direction.
•	In the original code, bars can only move in four directions if I understand it right (0 90 180 270). Users need to define start coordinates (x_start, y_start, x_end, y_end), speed in two orthogonal directions (x_delta, y_delta) and trial duration (frames).
•	After the change, Users need to define moving direction (direction), which can be arbitrary, bar width (bar_width) and speed (delta). Start coordinates will be determined by direction and bar width, trial duration will be set to the time it takes for the bar to traverse the whole display. These are implemented by line 98-154
•	A new function was added to /RSM_Utils :rotateData.m, which is called to calculate start coordinates.
o	Enable users to define interval between trials
•	Intervals are added to trial duration in line 138 and line 142

•	Moving_Grating.m
o	Add square-wave grating.
•	Users can specify ‘sine’ or ‘square’ in ‘subtype’, square-wave texture is generated in line 335-336
o	Change algorithm of the generation of grating texture 
•	Generate 1D texture instead of 2D grating: line 113-128, 334-344, 366.
•	Instead of generating a grating texture at the beginning of each loop, precompute a larger texture and display them one-by-one in the loop. Line 323-330, 353-356
•	The reason for making above two changes is to speed things up. In the original code, grating texture is downscaled by a factor of 4 (line 329), which is fine for sine wave grating but introduce blurring in square wave grating. However if the texture is not downscaled, the execution time for each loop is so long that makes the motion unsmooth.
•	Note: I didn’t change the ‘S_file constructor mode’ part because I didn’t figure out how to use this mode.
o	Enable users to define interval between trials
•	Line 386-390

•	PulseCombo.m
o	Add full field pulses stimulus (gray-white-gray-black)
•	Line 410-480, 699-766

•	Counterphase_Grating.m
o	Enable users to define interval between trials
•	Line 93-99, 407-411

•	Prerun_RNstim.m
o	Temporarily not to prerun raw movie stimulus
•	Prerun raw movie causes error. Comment line 48-71

•	Stim_Engine.m
o	Trigger
•	Temporarily move line 60-78 to the end of q_RSM.m. This is because we currently have the trigger issue that triggers from DAQ to stimulus computer are not working well. So we decided to only use the first trigger to initialize stimulus. And the subsequent triggers are actually unnecessary.
•	q_RSM.m
o	Add full field pulses
•	Line 73-75
o	Generate randomized stimulus sequence.
•	For MB, MG and CG. Save stimulus information in structure. Line 80-124
•	Add new function ‘rand_stim.m’ to /RSM_Utils
o	Calculate moving bar stimulus duration
•	Line 91-92
•	Add new function ‘calc_mb_duration.m’ to /RSM_Utils
o	Add ‘wait trigger’, ‘wait key’
•	See ‘Stim_Engine.m’ above. Line 149-160


2014-10-28: BV Modified Analyze_Monitor_Chan.m and Fit_PreProcced_RadPow_Curve.m to deal with problems raised by uncommenting the gun value normalization in Interpolate_Radpow_Data.m (see above entry for 2014-10-11). 


Stim_script.m
%% Initialization

% Tested on stimulus computer in Greg's lab. The only piece of code that
% does not work is the last one (Cone-Isolating Pulse; As setup by S-file)
% 
% Tested by Xiaoyang Yao
% 2015-04-02

Start_RSM;
mglSetGammaTable( RSM_GLOBAL.monitor.red_table, RSM_GLOBAL.monitor.green_table, RSM_GLOBAL.monitor.blue_table );

%% Stimulus
%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Focus Squares
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Focus squares. </strong>\n');
%
%-----------------------------------------------------------------------------
clear_pending_stim


stimulus = [];                                  % This clears the stimulus variable if it has been previously initialized.
stimulus.type = 'FS';                           % "type" field contains 2letter id code for directing RSM desired stimulus setup
stimulus.stim_width = 800;                      % "stim_width" sets width of focus square pattern in pixels. 
stimulus.stim_height = 600;                     % "stim_width" sets height of focus square pattern in pixels. 
stimulus.back_rgb = [0.5, 0.5, 0.5];            % "back_rgb" gives the color of the monitor 'background' (the focus square pattern sits in front of 
                                                % this background color. The range is normalized from 0 to 1. 
run_stimulus(display, stimulus);
clear stimulus;


%% Solid full-screen color
%------------------------------------------------------------------------------
%
fprintf('?\n\n<strong> Solid full-screen color. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus.type = 'SC';
stimulus.back_rgb = [0.5, 0.5, 0.5];              
stimulus.rgb = [1, 0, 0];                           % This sets the color of the solid full screen square. Range is 0-1.
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;   
stimulus.wait_trigger = 0;                          % "wait_trigger" is a logic flag. If set to 1 RSM will wait for a trigger event from the input line.
                                                    % This mode supercedes the wait_key mode. Thus, if 'wait_trigger = 1 and wait_key = 1' it is the 
                                                    % equivalent of 'wait_trigger = 1 and wait_key = 0'. 
stimulus.wait_key = 0;                              % "wait_key" is also a logic flag. wait_key = 1 prompts the user for hitting an input key to start the stimulus.
stimulus.x_start = 0;  stimulus.x_end = 800;        % These fields set the start and stop pixels that control stimulus size.
stimulus.y_start = 0;   stimulus.y_end = 600;

run_stimulus(display, stimulus);
clear stimulus;

%% Flashing full-screen color
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Flashing full-screen color. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'FC';  
stimulus.rgb = [1, 0, 0];
stimulus.back_rgb = [0.5, 0.5, 0.5];
stimulus.frames = 60;                       % "frames" is the number of frame refreshes to wait for each half-cycle (i.e. the pulse is on for the number of frames set here
                                            % and then off for the same number of frames. This completes one repetition of the pulse.
stimulus.num_reps = 3;                      % "num_reps" gives the number of times the pulse on-off cycle is completed.
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;

run_stimulus(display, stimulus);
clear stimulus;

%% Full Field Pulses (g-w-g-b)
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Full Field Pulses. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'FP';  
stimulus.rgb_white = [1, 1, 1];         % This sets the color of bright step
stimulus.rgb_black = [0, 0, 0];         % This sets the color of dark step
stimulus.back_rgb = [0.5, 0.5, 0.5];    % This sets the color of gray step
stimulus.frames = 60;                   % Number of frames for each step
stimulus.num_reps = 5;                  % Number of 'g-w-g-b' cycles
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
run_stimulus(display, stimulus);
clear stimulus;




%% Moving bar
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Moving bar. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'MB';
stimulus.num_reps = 1;   % This gives the number of times the bar moves across the screen within each individual trial.

stimulus.back_rgb = [0.5, 0.5, 0.5];
stimulus.rgb = [1, 1, 1];
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;
stimulus.bar_width = [120 240]; % pixel
stimulus.direction = [0 45 90 135 180 225 270 315];
stimulus.delta = [5];    % speed of the moving bar, unit: pixel/frame    
stimulus.interval = 0;   % duration of gray screen after each trial, unit: frame
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
stimulus.repeats = 1;    % This gives the number of cycles for each parameter combination

run_stimulus(display, stimulus);
clear stimulus;

%% Moving Grating
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Moving Grating. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim
stimulus = [];
stimulus.type = 'MG';
stimulus.subtype = 'square';     % 'square' or 'sine'
stimulus.back_rgb = [0, 0, 0];
stimulus.rgb = [1.0, 1.0, 1.0];
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;
stimulus.phase0 = 0;             % initial phase of the grating. Units are in radians.
stimulus.temporal_period = [2];  % sec
stimulus.spatial_period = [120]; % frame
stimulus.direction = [0 45 90 135 180 225 270 315];       % Convention 0 deg is 3 oclock
stimulus.frames = 240;           % number of frames for each stimulus trial.
stimulus.interval = 0;           % unit: second
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
stimulus.repeats = 2;


run_stimulus(display, stimulus);
clear stimulus;


%% Counterphase Grating
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Counterphase Grating. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'CG';  
stimulus.back_rgb = [0.0, 0.0, 0.0];
stimulus.rgb = [1.0, 1.0, 1.0];
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;
stimulus.phase0 = 0; 
stimulus.temporal_period = 1;  
stimulus.spatial_period = 60;
stimulus.direction = [0 45];

stimulus.frames = 300;        
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
stimulus.repeats = 1;
stimulus.interval = 2; %Sec
run_stimulus(display, stimulus);
clear stimulus;


%% Random Noise : Binary
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Random Noise : Binary. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'RN';

stimulus.back_rgb = [0.5, 0.5, 0.5];
stimulus.rgb = [1.0, 1.0, 1.0];
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;
stimulus.independent = 0;           % "independent = 1" denotes that each color's contribution is allowed to vary for a given stixel.
                                    % "independent = 0" denotes that each color's contribution is equal (i.e. stixel is grayscale).
stimulus.interval = 2;              % "interval" controls the number of frames that each random stimulus field is held before refresh.
stimulus.seed = 11111;              % The seed for the random number algorithm.
stimulus.x_start = 100;  stimulus.x_end = 700;  % These fields set the start and stop pixels for the overall stimulus field size.
stimulus.y_start = 0;   stimulus.y_end = 600;
stimulus.stixel_width = 15;      stimulus.stixel_height = 15;   % These fields given the size of each stixel in pixels    
stimulus.field_width = 40;       stimulus.field_height = 40;    % These fields control the number of stixels in each direction.    
stimulus.duration = 5;  % unit: second   
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
stimulus.interval_sync = 0;  % If "interval_sync" = 0 then no sync pulse square is shown. If this is set to a number (1-4) then a sync pulse square is displayed.
                             % Sync pulse setup is discussed in the next example.
stimulus.stop_frame = [];    % This field can be empty ("[]") in which case the stimulus proceeds till the duration time is reached.
                             % If this field is set to a frame number then the algorthim stops at that frame. (Very useful for debugging).


run_stimulus(display, stimulus);
clear stimulus;

%% Random Noise : Binary. Demo 2- sync pulse
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Random Noise : Binary. Demo 2- sync pulse. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'RN';
stimulus.back_rgb = [0.5, 0.5, 0.5];
stimulus.rgb = [1.0, 1.0, 1.0];
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;
stimulus.independent = 1;
stimulus.interval = 2;
stimulus.seed = 11111;
stimulus.x_start = 100;  stimulus.x_end = 700;
stimulus.y_start = 0;   stimulus.y_end = 600;
stimulus.stixel_width = 15;      stimulus.stixel_height = 15;       
stimulus.field_width = 40;        stimulus.field_height = 40;        
stimulus.duration = 12;     
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
stimulus.interval_sync = 0;
stimulus.stop_frame = [];
stimulus.interval_sync = 1;             % This sets the display of a sync pulse. 
                                        % The sync pulse square state changed between 'on' and 'off' when the stimulus array updates (every "interval" number of frames ).
                                        % 1 = Lower-left, 2 = Upper-left, 3 = Lower-Right, 4 = Upper-Right.
                                        
stimulus.interval_sync_xstart = 100;    % These fields define (in pixels) the location of the sync pulse on the screen.
stimulus.interval_sync_xend = 200;
stimulus.interval_sync_ystart = 100;
stimulus.interval_sync_yend = 200;


run_stimulus(display, stimulus);
clear stimulus;


%% Random Noise: Gaussian
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Random Noise: Gaussian. </strong>\n');
%
%------------------------------------------------------------------------------
% Note: Exept for two variables the Gaussian case is the same as the above examples of binary random noise. 
% First, the "independent" field is absent in the Gaussian case. This is because presently only gray-scale Gaussian stimuli are available. 
% Second, there is s new field "sigma" (described below). 

clear_pending_stim

stimulus = [];
stimulus.type = 'RG';

stimulus.back_rgb = [0.5, 0.5, 0.5];
stimulus.rgb = [1.0, 1.0, 1.0];
stimulus.rgb = stimulus.rgb - stimulus.back_rgb;
stimulus.interval = 3;
stimulus.seed = 11111;

stimulus.x_start = 100;  stimulus.x_end = 700;
stimulus.y_start = 0;   stimulus.y_end = 600;
stimulus.stixel_width = 15;      stimulus.stixel_height = 15;       
stimulus.field_width = 40;        stimulus.field_height = 40;        
stimulus.duration = 5;     
stimulus.wait_trigger = 0;
stimulus.wait_key = 0;
stimulus.interval_sync = 0;
stimulus.stop_frame = [];
stimulus.sigma = 0.16;     % This controls the 1-sigma width of the Gaussian gun command values in gun value units.


run_stimulus(display, stimulus);
clear stimulus;

%% Raw Movie, Demo 1- forward
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Raw Movie, Demo 1- forward. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'RM';                               % See above.
stimulus.fn = 'catcam_forest.rawMovie';             % The filename of the movie. Currently RSM assumes such movies are stored in RSM_Movies.
stimulus.back_rgb = [0.5, 0.5, 0.5];                % See above.
stimulus.x_cen_offset = 0;   stimulus.y_cen_offset = 100; % These fields control the centering of the movie stimulus in pixels. The size of the movie stimulus (in pixels) is given in the .rawMovie header information. 
                                                    % Positive x_cen_offset shifts the stimulus to 'screen-right'(ie the shift is
                                                    % toward the right side of the display monitor when seen face-on.
                                                    % Positive y_cen_offset shifts the stimulus downwards on the display monitor screen.
stimulus.interval = 1;                              % This is the same as in above examples: it gives the number of frames between stimulus changes. 
    % In the movie case there is something worthy of further comment: if
    % the interval = 1 and the display monitor refresh rate in Hz is higher than
    % the frame update rate of the recorded movie in Hz then the movie
    % appears speeded-up. 
stimulus.preload = 0;                               % This option is a hold-over feature but has potential future use for small movies. 
stimulus.wait_trigger = 0;                          % See above.
stimulus.wait_key = 1;                              % See above.
stimulus.interval_sync = 0;                         % See above.
stimulus.stop_frame = [];                           % See above.
stimulus.flip_flag = 1;                             % Movies can be presented with various flip options:
                                                    % 1 = normal; 2 = vertical flip (up-down reversal); 3 = horizontal flip (L-R reversal); 4 = vertical + horizontal flip
stimulus.reverse_flag = 0;                          % The default value of 0 allows the movie to be shown from frame 1 to last frame. 
                                                    % The reverse_flag =1 option is disussed below. 
stimulus.first_frame = 1;                           % This gives the frame number of the movie that is used as the lowest frame number of stimulus movie.
                                                    
stimulus.last_frame = [];                           % This gives the highest frame number of the stimulus movie. 
                                                    % If one wants to run to the end of a movie of unspecified length then "[]" is entered. 
                                                    % If set to '[]' then the final frame of the rawMovie (as stored in the raw file) is used as the highest frame number.  
 
run_stimulus(display, stimulus);
clear stimulus;

%% Raw Movie, Demo 2- reverse
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Raw Movie, Demo 2- reverse. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'RM';                               % See above.
stimulus.fn = 'catcam_forest.rawMovie';              % See above.
stimulus.back_rgb = [0.5, 0.5, 0.5];                 % See above.
stimulus.x_cen_offset = 0;   stimulus.y_cen_offset = 0; % See above.
stimulus.interval = 1;                              % See above.
stimulus.preload = 0;                               % See above.
stimulus.wait_trigger = 0;                          % See above.
stimulus.wait_key = 1;                              % See above.
stimulus.interval_sync = 0;                         % See above.
stimulus.stop_frame = [];                           % See above.
stimulus.flip_flag = 1;                             % See above.
stimulus.reverse_flag = 1;                          % Here we are showing the movie in reverse. 
                                                    % In this situation the last frame field controls where we start presentation.
                                                    % Frames are then presented sequentially in reverse order until the first frame entry is reached.
stimulus.first_frame = 1;                           % See above.
stimulus.last_frame = [];                           % See above.

run_stimulus(display, stimulus);
clear stimulus;

%% Raw Movie, Demo 3- reverse & upside down
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Raw Movie, Demo 3- reverse & upside down; frames 2000 down to 1000. </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];
stimulus.type = 'RM';                               % See above.
stimulus.fn = 'catcam_forest.rawMovie';             % See above.
stimulus.back_rgb = [0.5, 0.5, 0.5];                % See above.
stimulus.x_cen_offset = 0;   stimulus.y_cen_offset = 0; % See above.
stimulus.interval = 1;                              % See above.
stimulus.preload = 0;                               % See above.
stimulus.wait_trigger = 0;                          % See above.
stimulus.wait_key = 1;                              % See above.
stimulus.interval_sync = 0;                         % See above.
stimulus.stop_frame = [];                           % See above.
stimulus.flip_flag = 2;                             % In this example the movie is flipped upside-down.
stimulus.reverse_flag = 1;                          % We again reun in reverse order. 
stimulus.first_frame = 1000;                        % Since we are in reverse mode this gives the final frame of the presentation.
stimulus.last_frame = 2000;                         % This gives the beginning frame (since we are in reverse mode). 

run_stimulus(display, stimulus);
clear stimulus;

%% Cone-Isolating Pulse
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Cone-Isolating Pulse </strong>\n');
%
%------------------------------------------------------------------------------
clear_pending_stim

stimulus = [];                                      
stimulus.type = 'PL';                               % See above.
stimulus.map_file_name = '1234d01/map-0000.txt';    % The base path assumes that such files are held in the RSM_Map_Vault directory. The map_file_name gives the rest of the filename path within the RSM_Map_Vault. 
stimulus.lut_file_name = '1234d01/lut.mat';         % See above. 
stimulus.back_rgb = [0.5, 0.5, 0.5];                % See above.
stimulus.frames = 120;                              % "frames" is the number of frame refreshes to wait for each half-cycle (i.e. the pulse is on for the number of frames set here
                                                    % and then off for the same number of frames. This completes one repetition of the pulse.
stimulus.num_reps = 3;                              % See above.
stimulus.wait_trigger = 0;                          % See above.
stimulus.wait_key = 1;                              % See above.

run_stimulus(display, stimulus);
clear stimulus;

%% Cone-Isolating Pulse; As setup by S-file
%------------------------------------------------------------------------------
%
fprintf('\n\n<strong> Cone-Isolating Pulse; As setup by S-file. </strong>\n');
%
%-----------------------------------------------------------------------------
clear_pending_stim

fprintf('Test takes a long time to complete.\n');

run_test = Tested_Input_Logical('Enter [1] to run demo, or [0] to skip.');

if (run_test)

stimulus = [];
stimulus.type = 'PL';
stimulus.control_flag = 2;
stimulus.sfile_name = 's03';
stimulus.mapfile = '/1234d01/map-0000.txt';

run_stimulus(display, stimulus);

end

Stop_RSM




